src/test/java/com/squareup/javapoet/AnnotatedTypeNameTest.java

```diff
@@ -114,27 +117,81 @@ private void annotatedEquivalence(TypeName type) {
     assertNotEquals(type.hashCode(), type.annotated(NEVER_NULL).hashCode());
   }
 
-  // https://github.com/square/javapoet/issues/431
-  // @Target(ElementType.TYPE_USE) requires Java 1.8
+  private final static String TUA = TypeUseAnnotation.class.getCanonicalName();
+  private final AnnotationSpec TYPE_USE_ANNOTATION =
+      AnnotationSpec.builder(TypeUseAnnotation.class).build();
+
+  @Target(ElementType.TYPE_USE)
   public @interface TypeUseAnnotation {}
 
   // https://github.com/square/javapoet/issues/431
   @Ignore @Test public void annotatedNestedType() {
-    String expected = "java.util.Map.@" + TypeUseAnnotation.class.getCanonicalName() + " Entry";
-    AnnotationSpec typeUseAnnotation = AnnotationSpec.builder(TypeUseAnnotation.class).build();
-    TypeName type = TypeName.get(Map.Entry.class).annotated(typeUseAnnotation);
-    String actual = type.toString();
-    assertEquals(expected, actual);
+    TypeName type = TypeName.get(Map.Entry.class).annotated(TYPE_USE_ANNOTATION);
+    assertThat(type.toString()).isEqualTo("java.util.Map.@" + TUA + " Entry");
   }
 
   // https://github.com/square/javapoet/issues/431
   @Ignore @Test public void annotatedNestedParameterizedType() {
-    String expected = "java.util.Map.@" + TypeUseAnnotation.class.getCanonicalName()
-        + " Entry<java.lang.Byte, java.lang.Byte>";
-    AnnotationSpec typeUseAnnotation = AnnotationSpec.builder(TypeUseAnnotation.class).build();
     TypeName type = ParameterizedTypeName.get(Map.Entry.class, Byte.class, Byte.class)
-        .annotated(typeUseAnnotation);
-    String actual = type.toString();
-    assertEquals(expected, actual);
+        .annotated(TYPE_USE_ANNOTATION);
+    assertThat(type.toString())
+        .isEqualTo("java.util.Map.@" + TUA + " Entry<java.lang.Byte, java.lang.Byte>");
+  }
+
+  // https://github.com/square/javapoet/issues/614
+  @Ignore @Test public void annotatedArrayType() {
+    TypeName type = ArrayTypeName.of(ClassName.get(Object.class)).annotated(TYPE_USE_ANNOTATION);
+    assertThat(type.toString()).isEqualTo("java.lang.Object @" + TUA + " []");
```

Comment:
this one seems counterintuitive to me. The annotation is on the outermost type, but it’s printed here?

(not saying it’s incorrect, but wow, this feature is weird)

Reply:
I know.  See JLS [§9.7.4](https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html)

>Type annotations can apply to an array type or any component type thereof ([§10.1](https://docs.oracle.com/javase/specs/jls/se8/html/jls-10.html#jls-10.1)). For example, assuming that `A`, `B`, and `C` are annotation types meta-annotated with `@Target(ElementType.TYPE_USE)`, then given the field declaration:
>
>```java
>@C int @A [] @B [] f;
>```
>`@A` applies to the array type `int[][]`, `@B` applies to its component type `int[]`, and `@C` applies to the element type `int`. For more examples, see [§10.2](https://docs.oracle.com/javase/specs/jls/se8/html/jls-10.html#jls-10.2).

My guess is that this was chosen so that is consistent with array creation and access where the first pair of brackets refer to the outermost dimension. Still it does not feel right.